versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/IndexesLogicImpl.java

```diff
@@ -0,0 +1,536 @@
+              } else {
+                // purge old removes, if there is no reference-index that might contain those
+                c = null;
+              }
+              break;
+            case INCREMENTAL_REMOVE:
+              if (!hasReferenceIndex) {
+                // purge old removes, if there is no reference-index that might contain those
+                c = null;
+              }
+              break;
+            default:
+              break;
+          }
+          return c;
+        });
+    return i;
+  }
+
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  @Override
+  public Iterable<StoreIndexElement<CommitOp>> commitOperations(
+      @Nonnull @jakarta.annotation.Nonnull CommitObj commitObj) {
+    return commitOperations(incrementalIndexFromCommit(commitObj));
+  }
+
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  @Override
+  public Iterable<StoreIndexElement<CommitOp>> commitOperations(
+      @Nonnull @jakarta.annotation.Nonnull StoreIndex<CommitOp> index) {
+    return () ->
+        new AbstractIterator<StoreIndexElement<CommitOp>>() {
+          final Iterator<StoreIndexElement<CommitOp>> delegate = index.iterator();
+
+          @Override
+          protected StoreIndexElement<CommitOp> computeNext() {
+            Iterator<StoreIndexElement<CommitOp>> d = delegate;
+            while (true) {
+              if (!d.hasNext()) {
+                return endOfData();
+              }
+              StoreIndexElement<CommitOp> el = d.next();
+              if (el.content().action().currentCommit()) {
+                return el;
+              }
+            }
+          }
+        };
+  }
+
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  @Override
+  public StoreIndex<CommitOp> incrementalIndexFromCommit(
+      @Nonnull @jakarta.annotation.Nonnull CommitObj commit) {
+    return deserializeIndex(commit.incrementalIndex());
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public StoreIndex<CommitOp> buildCompleteIndex(
+      @Nonnull @jakarta.annotation.Nonnull CommitObj commit,
+      Optional<StoreIndex<CommitOp>> loadedIncrementalIndex) {
+    checkArgument(!commit.incompleteIndex(), "Commit %s has no complete key index", commit.id());
+
+    StoreIndex<CommitOp> incremental =
+        loadedIncrementalIndex.orElseGet(() -> incrementalIndexFromCommit(commit));
+    StoreIndex<CommitOp> index = incremental;
+
+    ObjId referenceIndexId = commit.referenceIndex();
+    List<IndexStripe> commitStripes = commit.referenceIndexStripes();
+    if (!commitStripes.isEmpty()) {
+      checkState(
+          referenceIndexId == null,
+          "Commit %s: must not have both pointer to a reference index and stripes",
+          commit.id());
+      StoreIndex<CommitOp> referenceIndex = referenceIndexFromStripes(commitStripes, commit.id());
+      index = layeredIndex(referenceIndex, incremental);
+    }
+    if (referenceIndexId != null) {
```

Comment:
Should this be `else if`? The `index` value from line 196 would be overwritten below.